### DISCUSSION TITLE
Setting up the API call with python

### DISCUSSION THREAD

**Discussion by orderlyBolt** [speechly/speechly#230] [Q&A] [answered]
Hi,

Can you provide an example of the API call to send a audio file and receive a transcription of the audio file?

**Comment by bigdatabaracus** [comment]
Hello @orderlyBolt, we have two api options: the streaming api and the batch api. 

To run the streaming api, use this example [here](https://github.com/speechly/api/blob/master/python/speechly_grpc_example.py) for Python. 

For the batch api you can find instructions and an example [here](https://docs.speechly.com/apis/batch-api/). The example is using `curl` but applying this to Python with something like the [requests](https://requests.readthedocs.io/en/latest/) library should rather straight forward. 

If you have a file locally that you want to run agains the batch api instead of something behind an url, instead of  `"uri": "https://docs.speechly.com/test1_en.wav"` you should do `"audio": "'$(base64<test1_en.wav)'"` 

Hope this helps! 

best regards, -Hannes-

**Reply by orderlyBolt** [reply]
Thank you for the fast reply! :D

However, i am getting unexpected tokens errors right now - can you elaborate one the correct format for uploading files. I am currently doing it as such

`import requests
import json

path = "test_file.wav"

headers = {
    'Context-Type': 'application/json',
    'Content-Type': 'application/x-www-form-urlencoded',
}

file = "'$(base64<" + path + ")'"

data = {"audio": file,
        "application": {  "appId":  "ffbca8ea-dc71-4573-832e-3da984627045" }}

response = requests.post('https://api.speechly.com/speechly.identity.v2.IdentityAPI/Login', 
                         headers=headers,
                         data=data,
                         verify = False)


response.text`

**Comment by langma** [comment, marked as answer]
Hi!

I created a quick example on how to call the Speechly BatchAPI using python `requests` library. There are comments between the sections that try to clarify what is done. Copy paste this, change the appId, and run :)

```python
import requests
import uuid
import base64
import time

# read an audio file in memory (note that the it should be PCM 16Khz 1 channels to get good results)
with open('test1_en.wav', 'rb') as f:
    audio_data = f.read()

# create a device ID (uuid)
deviceId = uuid.uuid4()

# get a Speechly access token to use the correct Speechly app
r = requests.post(
    'https://api.speechly.com/speechly.identity.v2.IdentityAPI/Login',
    json={'deviceId': str(deviceId), 'application': {'appId': 'YOUR_APP_ID'}}
)
token = r.json()['token']

# send the file to the BatchAPI to create a batch transcribe operation
batch_req = [{
    'config': {
        'encoding': 1,
        'channels': 1,
        'sampleRateHertz': 16000
    },
    'audio': base64.b64encode(audio_data).decode('ascii')
}]
r = requests.post(
    'https://api.speechly.com/speechly.slu.v1.BatchAPI/ProcessAudio',
    headers={'authorization':f'Bearer {token}'},
    json=batch_req
)
op = r.json()['operation']

# poll the BatchAPI, waiting for the batch operation to be done
while op['status'] != 'STATUS_DONE':
    time.sleep(1)
    r = requests.post(
        'https://api.speechly.com/speechly.slu.v1.BatchAPI/QueryStatus',
        headers={'authorization':f'Bearer {token}'},
        json={'id': op['id']}
    )
    op = r.json()['operation']
    if op['error'] != '':
        raise Exception('error in transcribe: ' + op['error'])

# collect the words from the transcripts
transcript = [w['word'] for w in op['transcripts']]
print(' '.join(transcript))
```
